### ISSUE TITLE
Semantics of function and function! in Thyme

### ISSUE THREAD

**Issue by milktrader** [JuliaStats/TimeSeries.jl#4]
The current implementation for `function` (i.e., `moving`, `upto`, `simple_return`, `log_return`, `lag`, `lead`) is to have it accept and return a `DataArray` whose length is retained, and `NAs` padded where it makes sense. 

`function!` accepts a `DataFrame` and mutates that same object with a new padded column. 

Going forward, I think it makes sense to leave the `function!` semantic alone. This would limit its use to `DataFrames` and that's an acceptable restriction. 

But I'd like to change the semantics for the non-bang version of `function` to accept either an `Array{T,1}` or a `DataArray` and return the same type object but without padding. Padding with `NAs` doesn't work with `Array` because it's only defined for `DataArray` and `DataFrame` and this limits this family of functions. 

To sum up:

`function` accepts a `Array{T,1}` or `DataArray` and returns the same type with a length equal to the original minus what was lost to computation (`lag` of length 1 would lose one row, for example)

`function!` only accepts `DataFrame` and returns that same `DataFrame` with a new column that is the result of the computation, and because of the requirements of `DataFrame` columns being of the same length, it pads with `NAs` where it makes sense to do so.
